packages/core/src/components/forms/_control-group.scss

```diff
@@ -76,69 +174,126 @@ Styleguide components.forms.control-group
     flex: 0 0 auto;
   }
 
+  // similarly to button groups, elements in control groups are stacked in a
+  // very particular order for best visual results. in each level of selector
+  // specificity, we define disabled styles last so that they override all other
+  // equally-specific styles (e.g. we don't want mouse interactions or focus
+  // changes to change the z-index of a disabled element).
+
   .pt-button,
   .pt-input,
-  .pt-select select {
+  .pt-select {
+    // create a new stacking context
+    position: relative;
+  }
+
+  .pt-input {
+    z-index: $control-group-z-index-input-default;
     // inherit radius since it's most likely to be zero
     border-radius: inherit;
 
     &:focus {
-      z-index: 1;
+      z-index: $control-group-z-index-input-focus;
+      border-radius: $pt-border-radius;
     }
-  }
 
-  &:not(.pt-vertical) > * {
```

Comment:
Moved this block to the bottom of the file to keep z-index-related stuff in one clump.